WorkflowSwiftUI/Sources/ObservableScreen.swift

```diff
@@ -0,0 +1,208 @@
+/// ``ObservableModel``, and then mapped to a screen implementation that uses that concrete model
+/// type. See ``ObservableModel`` for options on how to render one easily.
+public protocol ObservableScreen: Screen {
+    /// The type of the root view rendered by this screen.
+    associatedtype Content: View
+    /// The type of the model that this screen observes.
+    associatedtype Model: ObservableModel
+
+    /// The sizing options for the screen.
+    var sizingOptions: SwiftUIScreenSizingOptions { get }
+    /// The model that this screen observes.
+    var model: Model { get }
+
+    /// Constructs the root view for this screen. This is only called once to initialize the view.
+    /// After the initial construction, the view will be updated by injecting new values into the
+    /// store.
+    @ViewBuilder
+    static func makeView(store: Store<Model>) -> Content
+}
+
+public extension ObservableScreen {
+    var sizingOptions: SwiftUIScreenSizingOptions { [] }
+}
+
+public extension ObservableScreen {
+    func viewControllerDescription(environment: ViewEnvironment) -> ViewControllerDescription {
+        ViewControllerDescription(
+            type: ModeledHostingController<Model, Content>.self,
+            environment: environment,
+            build: {
+                let (store, setModel) = Store.make(model: model)
+                return ModeledHostingController(
+                    setModel: setModel,
+                    viewEnvironment: environment,
+                    rootView: Self.makeView(store: store),
+                    sizingOptions: sizingOptions
+                )
+            },
+            update: { hostingController in
+                hostingController.setModel(model)
+                hostingController.setViewEnvironment(environment)
+            }
+        )
+    }
+}
+
+public struct SwiftUIScreenSizingOptions: OptionSet {
+    public let rawValue: Int
+
+    public init(rawValue: Int) {
+        self.rawValue = rawValue
+    }
+
+    public static let preferredContentSize: SwiftUIScreenSizingOptions = .init(rawValue: 1 << 0)
+}
+
+private struct ViewEnvironmentModifier: ViewModifier {
+    @ObservedObject var holder: ViewEnvironmentHolder
+
+    func body(content: Content) -> some View {
+        content
+            .environment(\.viewEnvironment, holder.viewEnvironment)
+    }
+}
+
+private final class ViewEnvironmentHolder: ObservableObject {
+    @Published var viewEnvironment: ViewEnvironment
+
+    init(viewEnvironment: ViewEnvironment) {
+        self.viewEnvironment = viewEnvironment
+    }
+}
+
+private final class ModeledHostingController<Model, Content: View>: UIHostingController<ModifiedContent<Content, ViewEnvironmentModifier>> {
+    let setModel: (Model) -> Void
+    let setViewEnvironment: (ViewEnvironment) -> Void
+
+    var swiftUIScreenSizingOptions: SwiftUIScreenSizingOptions {
+        didSet {
```

Comment:
Should this have a `guard` on not equal?

Reply:
Got it – should it be, though? If I change the sizing param on a screen i'd expect it to apply